keyloader/keyloader.go

```diff
@@ -218,7 +218,65 @@ func configFactory() interface{} {
 ** CONSTRUCTORS
  */
 
+// NewKey returns a symmecrypt.Key object configured from a number of KeyConfig objects.
+// If several KeyConfigs are supplied, the returned Key will be composite.
+// A composite key encrypts with the latest Key (based on timestamp) and decrypts with any of they keys.
+// The KeyConfig parameters are expected to be sorted, with the latest key as first element.
+//
+// If the key configuration specifies it is sealed, the key returned will be wrapped by an unseal mechanism.
+// When the symmecrypt/seal global singleton gets unsealed, the key will become usable instantly. It will return errors in the meantime.
+//
+// The key cipher name is expected to match a KeyFactory that got registered through RegisterCipher().
+// Either use a built-in cipher, or make sure to register a proper factory for this cipher.
+// This KeyFactory will be called, either directly or when the symmecrypt/seal global singleton gets unsealed, if applicable.
+func NewKey(cfgs ...*KeyConfig) (symmecrypt.Key, error) {
+
+	if len(cfgs) == 0 {
+		return nil, errors.New("missing key config")
+	}
+
+	firstTS := cfgs[0].Timestamp
+	firstNonSealed := !cfgs[0].Sealed
+	comp := symmecrypt.CompositeKey{}
+
+	for _, cfg := range cfgs {
+
+		// ensure the first position was indeed the latest key (used for encryption in composite keys)
+		if cfg.Timestamp > firstTS {
+			return nil, errors.New("latest key is not in first position (key configs should be sorted)")
+		}
+		var ref symmecrypt.Key
+		factory, err := symmecrypt.GetKeyFactory(cfg.Cipher)
+		if err != nil {
+			return nil, err
+		}
+		if cfg.Sealed {
+			// if the first position (used for encryption in composite keys) was not sealed, but other keys used for fallback decryption are sealed
+			// it may be an attack to trigger a reencrypt with a key known by the attacker
+			if firstNonSealed {
+				return nil, errors.New("DANGER! Detected downgrade to non-sealed encryption key. Non-sealed key has higher priority, this looks malicious. Aborting!")
```

Comment:
If this is a danger, why not panic with this explicit message ?

Reply:
It did panic before, but that seems undesirable. Refusing to instantiate the Key mostly acomplishes the same thing as far as preserving the underlying data, I think, and it's neater overall.